### PR TITLE
Hook up pod template adapter

### PR DESCRIPTION
**Pull request by nlippis**
#### Release note

<hr>

##### Key changed/added classes in this PR
 * `KubernetesTaskRunnerFactory`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: docs/development/extensions-contrib/k8s-jobs.md [hunk 1]
```diff
@@ -28,7 +28,22 @@ Consider this an [EXPERIMENTAL](../experimental.md) feature mostly because it ha
...
+
+### Multi Container Pod Adapter
+The multi container pod adapter takes the podSpec of your `Overlord` pod and creates a kubernetes job from this podSpec.  It uses kubexit to manage dependency ordering between the main container that runs your druid peon and other sidecars defined in the `Overlord` pod spec.  To enable this pod adapter you can specify the runtime property `druid.indexer.runner.k8s.adapter.type: multiContainer` 
+
+### Pod Template Pod Adapter
+The pod template pod adapter allows you to specify a pod template file per task type.  This adapter requires you to specify a `base` pod spec which will be used in the case that a task specifc pod spec has not been defined.  To enable this pod adapter you can specify the runtime property `druid.indexer.runner.k8s.adapter.type: podAdapter`
+
+The base pod template must be specified as the runtime property `druid.indexer.runner.k8s.podTemplate.base: /path/to/basePodSpec.yaml`
+Task specific pod templates must be specified as the runtime property `druid.indexer.runner.k8s.podTemplate.{taskType}: /path/to/taskSpecificPodSpec.yaml` where {taskType} is the name of the task type i.e `index_parallel`
```
Comment (abhishekagarwal87) [inline docs/development/extensions-contrib/k8s-jobs.md:46]: I am wondering if there are better config values than what is proposed here. Specifically, `singleContainer` doesn't indicate that its based off Overlord pod spec. the pod template spec can be single container or multi container as well. I am struggling to come up with good suggestions myself. maybe something like `overlordSingleContainer`, `overlordMultiContainer`, `customTemplateAdapter`

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerConfig.java [hunk 2]
```diff
@@ -38,6 +38,7 @@
   @JsonProperty
   public boolean debugJobs = false;
 
+  @Deprecated
```
Comment (abhishekagarwal87) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerConfig.java:41]: can you add a comment about why this config is deprecated and the new alternative?

Comment (abhishekagarwal87) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerConfig.java:41]: there should be a change in docs too unless this property is not documented at all.

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerFactory.java [hunk 3]
```diff
@@ -84,18 +93,58 @@ public KubernetesTaskRunner build()
...
+    String adapter = properties.getProperty(String.format(
+        Locale.ROOT,
+        "%s.%s.adapter.type",
+        IndexingServiceModuleHelper.INDEXER_RUNNER_PROPERTY_PREFIX,
+        TYPE_NAME
+    ));
+
+    if (adapter != null && !MultiContainerTaskAdapter.TYPE.equals(adapter) && kubernetesTaskRunnerConfig.sidecarSupport) {
+      throw new IAE(
+          "Only kubernetes pod adapter [%s] can be specified when sidecarSupport is enabled",
```
Comment (abhishekagarwal87) [inline extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerFactory.java:123]: Please also add the value of `adapter` in this message.

Comment (nlippis) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerFactory.java:123]: The value of the adapter is interpolated here, check the next line that specifies `MultiContainerTaskAdapter.TYPE` as a format argument.

**Review by georgew5656 — APPROVED** [review]

File: extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerFactory.java (hunk 3 above)
Comment (abhishekagarwal87) [reply extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunnerFactory.java:123]: yes. that's the adapter that user has to specify. And then there is the adapter that user has currently specified. we should log both.

File: docs/development/extensions-contrib/k8s-jobs.md [hunk 4]
```diff
@@ -33,14 +33,14 @@ The K8s extension builds a pod spec using the specified pod adapter, the default
...
+The overlord multi container pod adapter takes the podSpec of your `Overlord` pod and creates a kubernetes job from this podSpec.  It uses kubexit to manage dependency ordering between the main container that runs your druid peon and other sidecars defined in the `Overlord` pod spec.  To enable this pod adapter you can specify the runtime property `druid.indexer.runner.k8s.adapter.type: overlordMultiContainer` 
 
-### Pod Template Pod Adapter
-The pod template pod adapter allows you to specify a pod template file per task type.  This adapter requires you to specify a `base` pod spec which will be used in the case that a task specifc pod spec has not been defined.  To enable this pod adapter you can specify the runtime property `druid.indexer.runner.k8s.adapter.type: podAdapter`
+### Custom Template Pod Adapter
+The pod template pod adapter allows you to specify a pod template file per task type.  This adapter requires you to specify a `base` pod spec which will be used in the case that a task specifc pod spec has not been defined.  To enable this pod adapter you can specify the runtime property `druid.indexer.runner.k8s.adapter.type: customTemplateAdapter`
```
Comment (abhishekagarwal87) [inline docs/development/extensions-contrib/k8s-jobs.md:43]:
```suggestion
The custom template pod adapter allows you to specify a pod template file per task type.  This adapter requires you to specify a `base` pod spec which will be used in the case that a task specifc pod spec has not been defined.  To enable this pod adapter you can specify the runtime property `druid.indexer.runner.k8s.adapter.type: customTemplateAdapter`
```

**Review by abhishekagarwal87 — APPROVED** [review]
LGTM assuming CI passes. Just one minor comment in the docs.

**Pull request merged**